File: lib/src/reconstruction/reconstruct_memory.dart
import '../graph/runtime_graph.dart';
import '../memory/runtime_memory.dart';
import '../memory/runtime_memory_graph.dart';
import 'reconstruct_graph.dart';

Map<String, dynamic> reconstructMemoryGraph(
  RuntimeGraph graph, [
  List<dynamic> history = const [],
]) {
  final memoryGraph = buildRuntimeMemoryGraph(graph, history);
  final fabric = buildRuntimeMemory(graph, history);
  return {
    'memory_graph': memoryGraph,
    'memory': fabric['memory'],
    'stable_hash': fabric['stable_hash'],
    'bounded': true,
  };
}

Map<String, dynamic> reconstructMemoryFromEnvelope(Map<String, dynamic> envelope) {
  final graph = reconstructRuntimeGraph(envelope);
  final history =
      ((envelope['runtime_memory'] as Map?)?['runtime_history'] as List?) ?? [];
  return reconstructMemoryGraph(graph, history);
}

File: src/reconstruction/reconstructMemory.ts
import { buildRuntimeMemory } from "../memory/runtimeMemory.js";
import { buildRuntimeMemoryGraph } from "../memory/runtimeMemoryGraph.js";
import type { RuntimeGraph } from "../contracts/graphContracts.js";

export function reconstructMemoryGraph(
  graph: RuntimeGraph,
  history: unknown[] = [],
): Record<string, unknown> {
  const memoryGraph = buildRuntimeMemoryGraph(graph, history);
  const fabric = buildRuntimeMemory(graph, history);
  return {
    memory_graph: memoryGraph,
    memory: fabric.memory,
    stable_hash: fabric.stable_hash,
    bounded: true,
  };
}

export function reconstructMemoryFromEnvelope(
  envelope: Record<string, unknown>,
): Record<string, unknown> {
  const graph = (envelope.unified_runtime_graph ?? { nodes: [], edges: [] }) as RuntimeGraph;
  const history =
    ((envelope.runtime_memory as Record<string, unknown> | undefined)?.runtime_history as unknown[]) ??
    [];
  return reconstructMemoryGraph(graph, history);
}
